dashboard/test/helpers/teacher_application_helper_test.rb

```diff
@@ -13,6 +13,23 @@ def sign_in(user)
     @incomplete_application = create TEACHER_APPLICATION_FACTORY, user: @user_with_two_incomplete_apps, status: 'incomplete'
     create TEACHER_APPLICATION_FACTORY, user: @user_with_two_incomplete_apps, status: 'incomplete', application_year: '2018-2019'
 
+    @user_with_incomplete_closed_app = create :teacher
+    regional_partner_with_closed_apps = create :regional_partner, apps_close_date_teacher: (Date.current - 3.days).strftime("%Y-%m-%d")
+    hash_with_rp_closed_apps = build :pd_teacher_application_hash, regional_partner_id: regional_partner_with_closed_apps.id
+    create TEACHER_APPLICATION_FACTORY,
+      user: @user_with_incomplete_closed_app,
+      status: 'incomplete',
+      regional_partner: regional_partner_with_closed_apps,
+      form_data_hash: hash_with_rp_closed_apps
+
+    @user_with_incomplete_open_app = create :teacher
+    regional_partner_with_open_apps = create :regional_partner, apps_close_date_teacher: (Date.current + 3.days).strftime("%Y-%m-%d")
```

Comment:
I didn't use Timecop here because I think three days is enough to avoid flakiness, and because more granular testing of times is happening in regional_partner_test

Reply:
yes, as long as 3 days is not the cutoff time (presumably even a close date of 1 second ago would be sufficient), this seems fine. if our tests take 3 days to run we've got other problems :-D 

Timecop is good to use by default in unit tests so that when you read the code (perhaps because the test starts failing) you don't have to think about whether there may be a timing bug. but I think you've got the variables named clearly enough in this case that it's clear you don't need Timecop.

Reply:
Helpful context! Since these are in the setup and we call the method that checks are_apps_closed inside each test, would I need to wrap a timecop in the whole test file? I was thinking that's a bit strange, but perhaps it's not a problem?

Reply:
Good question. I am not sure how Timecop.freeze works within the setup_all method. I don't see us using that pattern anywhere else. However I also don't really see a problem with having time frozen for the duration of the entire test file, so I think it could be worth a try :-)

Reply:
In theory it should be similar to using it in setup / teardown: https://github.com/code-dot-org/code-dot-org/blob/a6496fd65baca140f99b10a0c4fd09a50dd58c8e/dashboard/test/controllers/api/v1/user_school_infos_controller_test.rb#L4-L13

Reply:
Okay, I tried three different ways, none of which were successful 😢 . 
1) I first tried like the way you showed above with `teardown`. This was a bit illogical because I didn't use `teardown_all`, but the commit is here https://github.com/code-dot-org/code-dot-org/pull/51631/commits/cf943dcd521bfc8cc29c86e2297a00abc32d6b83
2) I then tried nesting the whole thing. I was hopeful for this one, but no dice: https://github.com/code-dot-org/code-dot-org/pull/51631/commits/d6f6f3b4c15c144302e27ebdafda32356892d12d
3) I went back to the `setup_all` and `teardown_all` approach, actually using `teardown_all` this time: https://github.com/code-dot-org/code-dot-org/pull/51631/commits/ea4321a2f9fba300892ecacdec33d347a7b8edba

In some ways, it feels a bit reassuring these don't work, because if we are running tests in parallel, freezing the time for an entire test file might mess things up? Although I'm guessing we don't run individual tests from the same file in parallel. Anyways, I'm going to merge it as-is without Timecop.

Reply:
yikes! ok, thanks for trying. I have very little experience with setup_all so I'm not sure what could be going on. I usually distrust it and start by moving things out of it and into setup whenever things are behaving strangely. anyway I think you've done your due diligence here. thanks for exploring :-)

Reply:
Sounds good! I feel good about using `setup_all` when creating variables that won't be updated further within the tests. Otherwise, I have no idea how it works 😬 . Might be good to keep in mind as we start refactoring the tests for the mailers.